### PR TITLE
Fixes for Hyper disconnect from the other end

### PR DESCRIPTION
**Pull request by tardyp**
Trying to fix issues we had last week on metabuildbot_travis, I had to fix several corner case issues on the core.

Each issue is fixed in a separate commit, while the integration tests I used to show them is in the last commit

### CODE REVIEW

File: master/buildbot/worker/latent.py [hunk 1]
```diff
@@ -210,7 +209,7 @@ def buildFinished(self, sb):
             if self.build_wait_timeout == 0:
                 # we insubstantiate asynchronously to trigger more bugs with
                 # the fake reactor
-                self.master.reactor.callLater(0, self.insubstantiate)
+                self.master.reactor.callLater(0, self._soft_disconnect)
```
Comment (seankelly) [inline master/buildbot/worker/latent.py]: :smile:

File: master/buildbot/test/util/integration.py [hunk 2]
```diff
@@ -182,7 +183,13 @@ def finishedCallback(_, data):
             yield self.master.data.updates.addChange(**useChange)
 
         # wait until we receive the build finished event
-        build = yield d
+        buildrequest = yield d
+        print(buildrequest)
```
Comment (seankelly) [inline master/buildbot/test/util/integration.py]: Should this print be here?

**Comment by seankelly** [comment]
Other than the errant `print`, things look good internally. When workers are disconnected, they are immediately available for the next build.
